cpp/bench/ann/src/raft/raft_ann_bench_utils.h

```diff
@@ -41,4 +41,72 @@ inline raft::distance::DistanceType parse_metric_type(raft::bench::ann::Metric m
     throw std::runtime_error("raft supports only metric type of inner product and L2");
   }
 }
+
+class configured_raft_resources {
+ public:
+  explicit configured_raft_resources(
+    const std::shared_ptr<rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>> mr)
+    : mr_{mr},
+      sync_{[]() {
+              auto* ev = new cudaEvent_t;
+              RAFT_CUDA_TRY(cudaEventCreate(ev, cudaEventDisableTiming));
+              return ev;
+            }(),
+            [](cudaEvent_t* ev) {
+              RAFT_CUDA_TRY_NO_THROW(cudaEventDestroy(*ev));
+              delete ev;
+            }},
+      res_{cudaStreamPerThread}
+  {
+  }
+
+  configured_raft_resources()
+    : configured_raft_resources{
+        {[]() {
+           auto* mr = new rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>{
+             rmm::mr::get_current_device_resource(), 1024 * 1024 * 1024ull};
+           rmm::mr::set_current_device_resource(mr);
+           return mr;
+         }(),
+         [](rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>* mr) {
+           if (rmm::mr::get_current_device_resource()->is_equal(*mr)) {
+             rmm::mr::set_current_device_resource(mr->get_upstream());
+           }
+           delete mr;
+         }}}
```

Comment:
suggestion: Initialize the members of the class by name, not as a struct.
suggestion: don't bury the re-setting of the previous device resource inside an unnamed deleter lambda. Just reset it in your class destructor.
question: does the mr_ really need to be shared? 
suggestion: use a unique_ptr.

```suggestion
  configured_raft_resources()
     : mr_{make_unique<rmm::mr::pool_memory_resource<rmm::mr::device_memory_resource>>(
             rmm::mr::get_current_device_resource(), 1024 * 1024 * 1024ull)}
{
}

~configured_raft_resources() 
{
  rmm::mr::set_current_device_resource(mr_->get_upstream());
}

```

Note that this also removes the call to `is_equal()` -- is there a reason in this benchmark that you are worried about the possibility of the current resource changing during the lifetime of this class?

In the future, please use `operator==` on MRs, because `is_equal` is likely to be removed.

Reply:
Hi Mark, thanks for having a look at this!

First of all, yes, we must use the shared pointer here. We create a new `configured_raft_resources` *at most* once per benchmark case, and let multiple concurrent threads spawned by gbench copy it. We do this because we need some of the resources not be shared due to thread safety (the raft handle and the event). However, we must to have exactly one rmm pool resource shared by everything, which is dictated by rmm current_device_resource semantics. Hence we need the pool resource be initialized with the original `configured_raft_resources` and destructed after the last copy of `configured_raft_resources` is destructed.

The same applies to setting/unsetting the rmm current_device_resource. Hence I cannot just reset it my class destructor and bury it inside an unnamed deleter lambda. An alternative would be to create a new class with the destructor that does the same and put it in the shared pointer, but why bother?

Also, I'm just curious as I'm not that experienced with the latest C++ best practices. What are the downsides of using curly braces to initialize the members "as a struct" as I did in this code?

Reply:
The downsides are that you aren't initializing them by name, so it is less readable, and if the order of the class members changes this initialization will be invalid. 

My suggestions were aimed at improving the readability and maintainability of the code, which currently I find obfuscated and error prone. But really this is not a library I have to read or maintain so it's up to you and the RAFT reviewers.

The one thing that I ask is that you remove the call to `is_equal()` for the resources, as this is probably going away.  If you must compare the resources, use `==`.

Reply:
See https://github.com/rapidsai/rmm/issues/1402

Reply:
@achirkin the big difference between the rmm example you've pointed out and the example here is that the rmm initializers are still naming the members so if the names should change or he rearranged, maintenance becomes much more straightforward.

I do agree with Mark here, and to be quite honest, I find the use of the pattern here (unnamed parameter initialization through lambdas) to be quite hard to read myself. I also prefer that we at least name the parameters, especially for a list that includes multiple different members. It seems like not much additional work in proportion to the significant gain in readability.

Reply:
Ok, now I really feel dumb that I don't see what you both mean :) I'm just calling here another constructor with a single argument that takes the pool, am I? If you're referring to the aggregate initialization, it's not even possible for this class, because it has two user-declared constructors.